sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala

```diff
@@ -139,33 +139,46 @@ object TableOutputResolver {
   private def resolveStructType(
       input: NamedExpression,
       inputType: StructType,
+      expectedCol: Attribute,
       expectedType: StructType,
-      expectedName: String,
       conf: SQLConf,
       addError: String => Unit,
       colPath: Seq[String]): Option[NamedExpression] = {
     val fields = inputType.zipWithIndex.map { case (f, i) =>
       Alias(GetStructField(input, i, Some(f.name)), f.name)()
     }
-    val reordered = reorderColumnsByName(fields, expectedType.toAttributes, conf, addError, colPath)
```

Comment:
Whenever we call `expectedType.toAttributes`, we ignore the raw type information associated with the attribute. As a result, we never added length checks when nested fields were resolved by name.